open-apparel-registry/open-apparel-registry
# Add dedupe-based facility list item matching

## Overview

Replaces the placeholder facility list matching process with a real, working process based on a trained [dedupe](https://github.com/dedupeio/dedupe) model.

Connects #221

## Notes

The provisioning time has increased a bit because dedupe depends on numpy which needs the `build-essential` apt package to be installed.

The initial version of this matcher used a pre-trained model downloaded from s3. Experiments with the pre-trained model revealed that the IDs of the canonical items used while training are returned as match results. This means that the model must be trained using actual data from the facilities table in order to have the proper IDs.

The `training.json` file that is read in during model training was created by running through an active learning session using the [oar_gazetteer_example.py](https://github.com/open-apparel-registry/open-apparel-registry/blob/develop/research/dedupe/gazetteer/oar_gazeteer_example.py) research script against the 42 example public facility CSVs used to populate the previous version of the application.

We have updated the Vagrantfile and docker-compose.yml file to ensure that our development environment is provisioned with 2 cores. This avoid a bug uncovered during our gazetteer research that only appears when there is a single core available. See dedupeio/dedupe#734

The new match process is designed to process the full list at once, and does so efficiently. We have refactored the `batch_process` code to separate "item" actions from "list" actions. "match" is the only a "list" action at the moment, and possibly ever, so we did not generalize the calling of the function that does the actual work. .

We moved the handling of the `parse` and `geocode` actions to a `process_items` method to keep the length of the `handle` method under control.

Since the matching process only works on an entire list at once we have also switched to using a normal job in AWS Batch for matching rather than an array job

We added the Unidecode PyPI package following the lead of the dedupe examples.

    Unidecode provides [a function that] takes Unicode data and tries to
    represent it in ASCII characters (i.e., the universally displayable
    characters between 0x00 and 0x7F), where the compromises taken when mapping
    between two character sets are chosen to be near what a human with a US
    keyboard would choose.

## Testing Instructions

* ./scripts/manage resetdb
* Register an account
* Upload  ⁨src⁩/django⁩/api⁩/management⁩/commands⁩/facility_lists⁩/13.csv
* Fully process the CSV
  * `./scripts/manage batch_process --list-id 15 --action parse`
  * `./scripts/manage batch_process --list-id 15 --action geocode`
  * `./scripts/manage batch_process --list-id 15 --action match`
* Browse the list http://localhost:6543/lists/15?page=1&rowsPerPage=100 and verify that the match process completed successfully and produced a mix of automatic matches, new facilities, and a pending match.
